pkg/maintainer/wallet/redemptions_test.go

```diff
@@ -37,3 +41,185 @@ func TestEstimateRedemptionFee(t *testing.T) {
+					nil,
+					&tbtc.NewWalletRegisteredEvent{
+						WalletPublicKeyHash: wallet.WalletPublicKeyHash,
+						BlockNumber:         wallet.RegistrationBlockNumber,
+					},
+				)
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+
+			// Record scenario pending redemptions to the local chain.
+			for _, pendingRedemption := range scenario.PendingRedemptions {
+				// Record the corresponding event. Set only relevant fields.
+				err = tbtcChain.AddPastRedemptionRequestedEvent(
+					&tbtc.RedemptionRequestedEventFilter{
+						// Remember about including the constant factor
+						// of 1000 blocks.
+						StartBlock:          scenario.ChainParameters.CurrentBlock - requestTimeoutBlocks - 1000,
+						WalletPublicKeyHash: [][20]byte{pendingRedemption.WalletPublicKeyHash},
+					},
+					&tbtc.RedemptionRequestedEvent{
+						WalletPublicKeyHash:  pendingRedemption.WalletPublicKeyHash,
+						RedeemerOutputScript: pendingRedemption.RedeemerOutputScript,
+						RequestedAmount:      pendingRedemption.RequestedAmount,
+						BlockNumber:          pendingRedemption.RequestBlock,
+					},
+				)
+
+				// Record the corresponding request object. Set only relevant fields.
+				tbtcChain.SetPendingRedemptionRequest(
+					pendingRedemption.WalletPublicKeyHash,
+					&tbtc.RedemptionRequest{
+						RedeemerOutputScript: pendingRedemption.RedeemerOutputScript,
+						RequestedAmount:      pendingRedemption.RequestedAmount,
+						RequestedAt:          pendingRedemption.RequestedAt,
+					},
+				)
+			}
+
+			walletsPendingRedemptions, err := walletmtr.FindPendingRedemptions(
+				tbtcChain,
+				scenario.Filter,
+			)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if diff := deep.Equal(
+				scenario.ExpectedWalletsPendingRedemptions,
+				walletsPendingRedemptions,
+			); diff != nil {
+				t.Errorf("invalid wallets pending redemptions: %v", diff)
+			}
+		})
+	}
+}
+
+func TestProposeRedemption(t *testing.T) {
+	fromHex := func(hexString string) []byte {
+		bytes, err := hex.DecodeString(hexString)
+		if err != nil {
+			t.Fatal(err)
+		}
+		return bytes
+	}
+
+	var walletPublicKeyHash [20]byte
+	copy(walletPublicKeyHash[:], fromHex(""))
+
+	redeemersOutputScripts := []bitcoin.Script{
+		fromHex("00140000000000000000000000000000000000000001"),
+		fromHex("00140000000000000000000000000000000000000002"),
+	}
+
+	var tests = map[string]struct {
+		fee              int64
+		expectedProposal *tbtc.RedemptionProposal
+	}{
+		"fee provided": {
+			fee: 10000,
+			expectedProposal: &tbtc.RedemptionProposal{
+				WalletPublicKeyHash:    walletPublicKeyHash,
+				RedeemersOutputScripts: redeemersOutputScripts,
+				RedemptionTxFee:        big.NewInt(10000),
+			},
+		},
+		"fee estimated": {
+			fee: 0, // trigger fee estimation
+			expectedProposal: &tbtc.RedemptionProposal{
+				WalletPublicKeyHash:    walletPublicKeyHash,
+				RedeemersOutputScripts: redeemersOutputScripts,
+				RedemptionTxFee:        big.NewInt(4300),
+			},
+		},
```

Comment:
We could also add a test case proving the proposal will not pass if the validation failed.